chia/util/cached_bls.py

```diff
@@ -0,0 +1,49 @@
+import functools
+from typing import List, Optional
+
+from blspy import AugSchemeMPL, G1Element, G2Element, GTElement
+from chia.util.hash import std_hash
+from chia.util.lru_cache import LRUCache
+
+
+def get_pairings(cache: LRUCache, pks: List[G1Element], msgs: List[bytes], force: bool) -> List[GTElement]:
+    pairings: List[GTElement] = []
+    missing_count: int = 0
+    for pk, msg in zip(pks, msgs):
+        aug_msg: bytes = bytes(pk) + msg
+        h: bytes = bytes(std_hash(aug_msg))
+        pairing: Optional[GTElement] = cache.get(h)
+        if not force and pairing is None:
+            missing_count += 1
+            # Heuristic to avoid more expensive sig validation with pairing
+            # cache when it's empty and cached pairings won't be useful later
+            # (e.g. while syncing)
+            if missing_count > len(pks) // 2:
+                return []
+        pairings.append(pairing)
+
+    for i, pairing in enumerate(pairings):
+        if pairing is None:
+            aug_msg = bytes(pks[i]) + msgs[i]
+            aug_hash: G2Element = AugSchemeMPL.g2_from_message(aug_msg)
+            pairing = pks[i].pair(aug_hash)
+
+            h = bytes(std_hash(aug_msg))
+            cache.put(h, pairing)
+            pairings[i] = pairing
+
+    return pairings
+
+
+local_cache = LRUCache(10000)
+
+
+def aggregate_verify(pks: List[G1Element], msgs: List[bytes], sig: G2Element, force: bool = False):
+    if len(pks) == 0 or len(msgs) == 0:
```

Comment:
I'm not sure these short-circuit cases are necessary since they would be fast anyway, and it's unlikely they'd ever be used.

Reply:
The special case with zero messages is actually relevant (e.g. a block with no transactions), but I removed this code anyway since it's correctly handled under `if len(pairings) == 0`.